agent/cache/watch.go

```diff
@@ -57,66 +64,135 @@ func (c *Cache) Notify(ctx context.Context, t string, r Request,
 	if !ok {
 		return fmt.Errorf("unknown type in cache: %s", t)
 	}
-	if !tEntry.Type.SupportsBlocking() {
-		return fmt.Errorf("watch requires the type to support blocking")
+	if tEntry.Type.SupportsBlocking() {
+		go c.notifyBlockingQuery(ctx, t, r, correlationID, ch)
+	} else {
+		info := r.CacheInfo()
+		if info.MaxAge == 0 {
+			return fmt.Errorf("Cannot use Notify for polling cache types without specifying the MaxAge")
```

Comment:
Right now I am using the MaxAge value as the amount of time to wait between internal getWithIndex calls. If this is 0, I could use the default interval but it will still come back with any data that happens to be in the cache.

One original implementation I had for this had a separate NotifyPolling function with the interval passed in. Then I realized that my upstreams were never getting updated. I dug a little further and found that the MaxAge was keeping any further fetches from happening. That isn't really what the function is meant to do so I changed it. The other idea was to detect MaxAge of 0 and set it to the default (which was why that default existed). However I didn't like the idea of just modifying the users data.

In the end calling this function without specifying the MaxAge in the request doesn't really make sense and it really would be a programming error not a runtime error. I was even considering making that fmt.Errorf a panic instead as if we ever get into that scenario then we haven't coded up the caller executing the notify properly.

Reply:
Yeah I should have come back here - the semantics of MaxAge were discussed elsewhere but I see the need for it now.